crates/chain/src/pow_chain.rs

```diff
@@ -0,0 +1,574 @@
+        chain.apply_changeset(&changeset)?;
+
+        debug_assert!(chain._check_changeset_is_applied(&changeset));
+
+        Ok(chain)
+    }
+
+    /// Initialize [`PoWChain`] with specified data.
+    pub fn from_data(
+        input: impl Iterator<Item = (u32, BlockHash, Target)>,
+        genesis_header: Header,
+    ) -> Result<Self, std::boxed::Box<dyn std::error::Error>> {
+        let tip = CheckPoint::<Header>::from_data(0, genesis_header);
+
+        let mut difficulty_map = HashMap::new();
+        let mut trusted_blocks = Vec::new();
+
+        for (height, hash, target) in input {
+            difficulty_map.insert(hash, Target::from_hex(&format!("{:x}", target))?);
+            trusted_blocks.push(BlockId { height, hash });
+        }
+
+        Ok(PoWChain {
+            difficulty_map,
+            trusted_blocks,
+            all_tips: vec![tip.clone()],
+            tip,
+        })
+    }
+
+    /// Load known checkpoints from external `checkpoints.json` file.
+    pub fn from_electrum_json(
+        &self,
+        path: &str,
+        genesis_header: Header,
+    ) -> Result<Self, std::boxed::Box<dyn std::error::Error>> {
+        let tip = CheckPoint::<Header>::from_data(0, genesis_header);
+
+        let mut difficulty_map = HashMap::new();
+        let mut trusted_blocks = Vec::new();
+
+        // Populate a vector of (blockhash, difficulty target) from external `checkpoints.json`
+        // file.
+        let file = std::fs::File::open(path)?;
+        let checkpoint_entries: Vec<(BlockHash, u64)> = serde_json::from_reader(file)?;
+
+        // Calculate heights based on the starting height of first block in `checkpoints.json`,
+        // which is block 2015. Subsequent blocks have height gaps of 2016, which is the number of
+        // blocks between each difficulty adjustment.
+        let mut height = 2015;
+
+        for (hash, target) in checkpoint_entries {
+            difficulty_map.insert(hash, Target::from_hex(&format!("{:x}", target))?);
+            trusted_blocks.push(BlockId { height, hash });
+
+            // Increment height by 2016, which is the number of blocks between each difficulty
+            // adjustment.
+            height += 2016;
+        }
+
+        Ok(PoWChain {
+            difficulty_map,
+            trusted_blocks,
+            all_tips: vec![tip.clone()],
+            tip,
+        })
+    }
+
+    /// Get the highest checkpoint.
+    pub fn tip(&self) -> CheckPoint<Header> {
+        self.tip.clone()
+    }
+
+    /// Apply the given `changeset`.
+    pub fn apply_changeset(&mut self, changeset: &ChangeSet) -> Result<(), MissingGenesisError> {
+        let old_tip = self.tip.clone();
+        let new_tip =
+            apply_changeset_to_checkpoint(old_tip, changeset, self.difficulty_map.clone())?;
+        self.tip = new_tip;
+        debug_assert!(self._check_changeset_is_applied(changeset));
+        Ok(())
+    }
```

Comment:
I'm finding this quite hard to reason about and it seems rather dangerous. How do you envision it being used here?